pkg/test/framework/components/echo/echotest/filters_test.go

```diff
@@ -249,3 +202,98 @@ func compare(t *testing.T, got echo.Instances, want echo.Instances) {
 		t.Errorf("did not include %v", expected)
 	}
 }
+
+func TestRun(t *testing.T) {
+	// source svc/cluster -> dest services -> number of subtests (should == num clusters)
+	framework.NewTest(t).Run(func(t framework.TestContext) {
+		tests := map[string]struct {
+			run    func(t framework.TestContext, testTopology map[string]map[string]int)
+			expect map[string]map[string]int
+		}{
+			"Run_WithDefaultFilters": {
+				run: func(t framework.TestContext, testTopology map[string]map[string]int) {
+					New(t, all).
+						WithDefaultFilters().
+						Run(func(ctx framework.TestContext, src echo.Instance, dst echo.Instances) {
+							// TODO if the destinations would change based on which cluster then add cluster to srCkey
+							srcKey := src.Config().FQDN()
+							dstKey := dst[0].Config().FQDN()
+							if testTopology[srcKey] == nil {
+								testTopology[srcKey] = map[string]int{}
+							}
+							testTopology[srcKey][dstKey]++
+						})
+				},
+				expect: map[string]map[string]int{
+					"a.echo.svc.cluster.local": {
+						"b.echo.svc.cluster.local":        2,
+						"external.echo.svc.cluster.local": 2,
+						"headless.echo.svc.cluster.local": 2,
+						"naked.echo.svc.cluster.local":    2,
+						"vm.echo.svc.cluster.local":       2,
+					},
+					"headless.echo.svc.cluster.local": {
+						"b.echo.svc.cluster.local":        2,
+						"external.echo.svc.cluster.local": 2,
+						"headless.echo.svc.cluster.local": 2,
+						"naked.echo.svc.cluster.local":    2,
+						"vm.echo.svc.cluster.local":       2,
+					},
+					"naked.echo.svc.cluster.local": {
+						"b.echo.svc.cluster.local":        2,
+						"external.echo.svc.cluster.local": 2,
+						"headless.echo.svc.cluster.local": 2,
+						"naked.echo.svc.cluster.local":    2,
+					},
+					"vm.echo.svc.cluster.local": {
+						"b.echo.svc.cluster.local":        2,
+						"headless.echo.svc.cluster.local": 2,
+						"naked.echo.svc.cluster.local":    2,
+						"vm.echo.svc.cluster.local":       2,
+					},
+				},
+			},
+			"RunToN": {
+				run: func(t framework.TestContext, testTopology map[string]map[string]int) {
+					noNaked := Not(MatcherAsFilter(echo.IsNaked()))
+					noHeadless := Not(MatcherAsFilter(echo.IsHeadless()))
+					New(t, all).
+						WithDefaultFilters().
+						From(noNaked, noHeadless).
+						To(noHeadless, Not(ExternalServices)).
+						RunToN(3, func(ctx framework.TestContext, src echo.Instance, dsts []echo.Instances) {
+							srcKey := src.Config().FQDN()
+							if testTopology[srcKey] == nil {
+								testTopology[srcKey] = map[string]int{}
+							}
+							var dstnames []string
+							for _, dst := range dsts {
+								dstnames = append(dstnames, dst[0].Config().FQDN())
+							}
+							dstKey := strings.Join(dstnames, "_")
+							testTopology[srcKey][dstKey]++
+						})
+				},
+				expect: map[string]map[string]int{
```

Comment:
do the following cases seem reasonable for traffic shifting? that's what I based this meta-test case on. 

cc @howardjohn

Reply:
Yes.

If we had external services in the picture what would it look like? Would we get different combinations of 3?

Reply:
Updated this to include external, the implementation handled that poorly (and completely dropped VM as a client).